### ISSUE TITLE
Problem with st_area()

### ISSUE THREAD

**Issue by hhmacedo** [r-spatial/sf#2146]
I have an area on the map with more than a million squared meters but when I use `st_area()` the result is less than 1 squared meter.

**To Reproduce**
Get [this shape](https://geoftp.ibge.gov.br/organizacao_do_territorio/malhas_territoriais/malhas_de_setores_censitarios__divisoes_intramunicipais/censo_2010/setores_censitarios_shp/df/df_setores_censitarios.zip) file from 2010 em [this](https://geoftp.ibge.gov.br/organizacao_do_territorio/malhas_territoriais/malhas_de_setores_censitarios__divisoes_intramunicipais/2021/Malha_de_setores_(shp)_por_UFs/DF/DF_Setores_2021.zip) one from 2021. 

```
library(dplyr)
library(sf)

shape_2010 <- read_sf(file.path("shapes", "53SEE250GC_SIR.shp"),
                               options = "ENCODING=WINDOWS-1252")

shape_2021 <- read_sf(file.path("shapes", "DF_Setores_2021.shp"))
```
Here is the result from `st_area()`:
```
> st_area(
+   st_intersection(
+     filter(shape_2010, CD_GEOCODI == "530010805060262")$geometry,
+     filter(shape_2021, CD_SETOR == "530010805060262")$geometry
+   )
+ )
Units: [m^2]
[1]       0 1338687
```
I've got almost the same result using `st_geod_area()`:
```
> st_geod_area(
+   st_intersection(
+     pull(filter(shape_2010, CD_GEOCODI == "530010805060262"), geometry),
+     pull(filter(shape_2021, CD_SETOR == "530010805060262"), geometry)
+   )
+ )
Units: [m^2]
[1]       0 1334014
```

<details>

```
> sessionInfo()
R version 4.2.2 (2022-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=Portuguese_Brazil.utf8  LC_CTYPE=Portuguese_Brazil.utf8   
[3] LC_MONETARY=Portuguese_Brazil.utf8 LC_NUMERIC=C                      
[5] LC_TIME=Portuguese_Brazil.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lwgeom_0.2-11 purrr_1.0.1   units_0.8-1   sf_1.0-12     ggplot2_3.4.1
[6] dplyr_1.1.0  

loaded via a namespace (and not attached):
 [1] zip_2.2.2          Rcpp_1.0.10        pillar_1.8.1       compiler_4.2.2    
 [5] class_7.3-20       tools_4.2.2        digest_0.6.31      lifecycle_1.0.3   
 [9] tibble_3.2.0       gtable_0.3.2       pkgconfig_2.0.3    rlang_1.1.0       
[13] DBI_1.1.3          cli_3.6.0          rstudioapi_0.14    parallel_4.2.2    
[17] e1071_1.7-13       furrr_0.3.1        withr_2.5.0        s2_1.1.2          
[21] globals_0.16.2     generics_0.1.3     vctrs_0.6.0        classInt_0.4-9    
[25] grid_4.2.2         tidyselect_1.2.0   glue_1.6.2         listenv_0.9.0     
[29] R6_2.5.1           parallelly_1.35.0  fansi_1.0.4        tidyr_1.3.0       
[33] farver_2.1.1       magrittr_2.0.3     codetools_0.2-18   scales_1.2.1      
[37] future_1.32.0      colorspace_2.1-0   utf8_1.2.3         KernSmooth_2.23-20
[41] proxy_0.4-27       wk_0.7.1           munsell_0.5.0      crayon_1.5.2      
```

```
> sf::sf_extSoftVersion()
          GEOS           GDAL         proj.4 GDAL_with_GEOS     USE_PROJ_H 
       "3.9.3"        "3.5.2"        "8.2.1"         "true"         "true" 
          PROJ 
       "8.2.1"
```
</details>

**Comment by kadyb** [comment]
I see that as a result of the intersection you got two objects -- the first linestring with area of 0 m^2 and the second polygon with area of 1338687 m^2. If you don't want linestring then add `dimensions = "polygon"` argument to `st_intersection()`.

**Comment by edzer** [comment]
We should document that.

**Comment by kadyb** [comment]
From what I remember, a similar issue (different intersection results in GEOS and S2) was raised several months ago by @jlacko, but I don't remember the details.

**Comment by jlacko** [comment]
If I remember correctly the issue in question was #2005 - and it involved areal weighted interpolation, and it is now closed via #2011 

The solution involved specifying `dimensions` argument for `sf::st_intersection()` call used under the hood of `sf::st_interpolate_aw()` via the dot dot dot argument - it affects only s2 calculation (in GEOS mode it is safely ignored). It tells s2 to ignore point and linestring geometries, as these have no defined area.

Which is exactly what you @kadyb is suggesting in the current case.
